modules/cco-ccoctl-upgrading.adoc

```diff
@@ -0,0 +1,89 @@
+====
+
+.Prerequisites
+
+* Obtain the {product-title} release image for the version that you are upgrading to.
+
+* Extract and prepare the updated `ccoctl` binary from the release image.
+
+//The test plan has the steps as extract the CredentialsRequest CRs, then extract the ccoctl binary. Does it work to include "Extract and prepare the updated `ccoctl` binary from the release image" as a prereq instead? Makes the task more similar to the install task.
+
+.Procedure
+
+. Extract the list of `CredentialsRequest` objects from the {product-title} release image:
++
+[source,terminal,subs="+quotes"]
+----
+$ oc adm release extract --credentials-requests --cloud=aws --to=__<path_to_directory_with_list_of_credentials_requests>__/credrequests quay.io/__<path_to>__/ocp-release:__<version>__
+----
+
+. Complete appropriate health checks. #elaborate#
+//Test plan has a link to several health checks, but what specifically should the user run and what are they watching out for?
+
+. Use the `ccoctl` tool to process all `CredentialsRequest` objects in the `credrequests` directory:
++
+[source,terminal,subs="+quotes"]
+----
+$ ccoctl aws create-iam-roles --name __<name>__ --region=__<aws_region>__ --credentials-requests-dir=__<path_to_directory_with_list_of_credentials_requests>__/credrequests --identity-provider-arn arn:aws:iam::__<aws_account_id>__:oidc-provider/__<name>__-oidc.s3.__<aws_region>__.amazonaws.com
+----
++
+where:
++
+** _<name>_ is the name used to tag any cloud resources that are created for tracking.
+** _<aws_account_id>_ is the AWS account ID.
+** _<aws_region>_ is the AWS region in which cloud resources will be created.
++
+[NOTE]
+====
+For AWS environments that use alternative IAM API endpoints, such as GovCloud, you must also specify your region with the `--region` parameter.
+====
++
+For each `CredentialsRequest` object, `ccoctl` creates an IAM role with a trust policy that is tied to the specified OIDC identity provider, and a permissions policy as defined in each `CredentialsRequest` object from the {product-title} release image.
+
+. Add any new permissions in the appropriate role and policy on AWS.
+//At this point in the test plan, it is stated that "there are 2 new permissions added for machine-api." How does the user figure this out? Would output from the previous step help?
+//Not really clear on the correct thing to tell the user to do on the AWS side, what is this process called? Any AWS docs we should refer them to?
++
+It may take up to several minutes for the updated permissions to take effect.
+
+. Complete appropriate health checks. #elaborate#
+//Again, test plan has a link to several health checks, but what specifically should the user run and what are they watching out for?
+
+.Verification
+
+//Does this make sense for the upgrade as well?
```

Comment:
I don't know that viewing these files helps. These files are typically passed to the installer, and the installer will install/apply them. For the upgrade case, we simply need to apply all the files that end in "credentials.yaml".
So something like:
ls _<path_to_ccoctl_output_dir>_/manifests/*-credentials.yaml | xargs -I{} oc apply -f {}

What may trip someone up is if there is a new CredentialsRequest that wants the Secret placed in a namespace that does not yet exist (it would be created as part of the actual upgrade process). The above command will fail to create those Secrets because of the missing namespace, and you would need to `oc create namespace <name of missing namespace>` for each Secret that failed to apply because of non-existent namespace and re-run the above command.

At this point, the IAM Roles are update, any new IAM Roles have been created, and we've applied the Secrets for the new component(s) (if there were new CredentialsRequests in this specific release image).

Reply:
On the upgrade case, we get no help. We have to do what the installer would do for us. Which is basically just taking those yaml files with Secret data in them, and just applying them once the control plane is up and running.
So for upgrade, we:
1) generate the IAM Roles (from the list of extracted CredentialsRequests)
2) generate the OpenShift Secrets (that we normally feed to the installer)
3) apply the generated Secrets to the live cluster

During install, some other pieces are responsible for creating the namespaces, but since we're basically doing some pre-upgrade steps, we need to create any namespaces for any Secrets that we generated (so that when the new component that will run in this new namespace is installed, it will have the credentials it needs immediately). The only times you would need to create a namespace is if there was a brand new component in this new version of OpenShift, and because we haven't yet upgraded, the new namespace isn't there yet. There is no fancy "create this secret and any namespace if necessary" command that I'm aware of, so you end up having to perform the two steps explicitly:
1) create the namespace to hold the Secret
2) create the Secret

Reply:
And yes, we can leave the "verify the IAM roles created by querying AWS.." bit. it is harmless.